Bvh animation

spodila1 opened the pull request:
Loads BVH Animation and plays on avatar. 

Tested with gvr-avatar demo with the following assets

Model -- YBOT
BvhAnimations -- Samba_Dancing_mixamo.com.bvh, Zombie_Stand_Up_mixamo.com, Football_Hike_mixamo.com

https://github.com/gearvrf/GearVRf-Demos/pull/641

GearVRf-DCO-1.0-Signed-off-by: Sahithi Podila sah.pod@partner.samsung.com

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRShader.java:
```diff
@@ -51,7 +51,7 @@
 public class GVRShader
 {
-    protected boolean mWriteShadersToDisk = false;
+    protected boolean mWriteShadersToDisk = true;
```
Please set back to false.

thomasflynn replied:
this should probably be false, right?

liaxim commented:
Please revert the formatting-only changes to jassimp.cpp.

NolaDonato commented:
I have also tested it. I have a few changes I would like to add to support importing the skeleton from the BVH file.

liaxim commented:
The updated gvr-avatar doesn't work for me.

spodila1 commented:
I have not added any code changes in jassimp.cpp. 

I tested with gvr-avatar demo (Note8). It is working fine.

NolaDonato commented:
I made some minor modifications to the BVH parser so it could handle mixed spaces and tabs. I also exposed the ability to read the skeleton independently from the motion.

I also pushed some of the skeleton and pose data from Java to C++ so physics C++ code can access it. This code will be tested by the retargeting demo I am working on.

NolaDonato commented:
I also pushed some minor changes to the gvr-avatar demo. It works for me with the YBot character and its BVH animations.

liaxim commented:
Please test with more demo apps and let us know which. Thanks.

mWriteShadersToDisk is still true. jassimp.cpp still has a ton of formatting-only changes.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimation.java:
```diff
@@ -393,6 +439,12 @@ final boolean onDrawFrame(float frameTime) {
         return stillRunning;
     }
 
+    public void animate(float timeInSec)
```
Seems like it doesn't have to be public. Please remove the qualifier or add javadoc. Preferably remove the qualifier.

NolaDonato commented:
?Actually applications can call it to step the animation programmatically


________________________________
From: Mihail <notifications@github.com>
Sent: Thursday, September 20, 2018 12:48 PM
To: Samsung/GearVRf
Cc: Nola Donato; Comment
Subject: Re: [Samsung/GearVRf] Bvh animation (#1987)


@liaxim commented on this pull request.

________________________________

In GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimation.java<https://github.com/Samsung/GearVRf/pull/1987#discussion_r219295242>:

> @@ -393,6 +439,12 @@ final boolean onDrawFrame(float frameTime) {
         return stillRunning;
     }

+    public void animate(float timeInSec)


Seems like it doesn't have to be public. Please remove the qualifier or add javadoc. Preferably remove the qualifier.

-
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/Samsung/GearVRf/pull/1987#pullrequestreview-157438359>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQw1S9ndsn1VXLWP4Qvlsd-kLPHTKj05ks5uc_EWgaJpZM4WvR7t>.

NolaDonato commented:
The javadocs are added and a few bugs are fixed. Ready to go. Seems to run the demo OK on my S8.

liaxim commented:
It is great that gvr-avatar works but a meaningful subset of the demo apps have to be tested. Some using animations are already broken so I dunno about them but anything with physics should be sanity checked. Let me know which demo apps have been verified. Thanks.

NolaDonato commented:
I tested with immersivepedia, meshanimation, controls and morph

liaxim commented on GVRf/Extensions/gvrf-physics/src/main/java/org/gearvrf/physics/GVRPhysicsLoader.java:
```diff
@@ -46,7 +46,8 @@
      * @param scene The scene containing the objects to attach physics components.
      */
-    public static void loadPhysicsFile(GVRContext gvrContext, String fileName, GVRScene scene) {
+    public static void loadPhysicsFile(GVRContext gvrContext, String fileName, GVRScene scene) throws IOException
```
gvr-physicsload doesn't compile:

```
GearVRf-Demos/gvr-physicsload/app/src/main/java/org/gearvrf/physicsload/MainActivity.java:194: error: unreported exception IOException; must be caught or declared to be thrown
            GVRPhysicsLoader.loadPhysicsFile(gvrContext, "fixed_slider.bullet", mainScene);
                                            ^
1 error
```

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimator.java:
```diff
@@ -202,6 +218,35 @@ public void setRepeatMode(int repeatMode)
             anim.setRepeatMode(repeatMode);
         }
     }
+    public void setOffset(float startOffset)
```
Bunch of new public APIs need javadoc.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAvatar.java:
```diff
@@ -9,14 +9,13 @@
 import org.gearvrf.GVRResourceVolume;
```
Needs copyright header.

liaxim commented:
gvr-cubemap doesn't work for me with this PR. On tap I get a black screen.

liaxim commented:
gvr-morph, gvr-outline, gvr-videoplayer - slower to load
gvr-cubemap - not all cubemaps show up
gvr-physicsload - doesn't compile
gvr-widgetviewer - some models are all black

Is the avatar supposed to tremble like that?

NolaDonato commented:
Yes, the avatar trembles because these are videos recorded from a body tracking setup. The original BVH files include the trembling.

NolaDonato commented:
Yes, good catch

thomasflynn commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimation.java:
```diff
@@ -332,13 +388,16 @@ public GVRAnimation start(GVRAnimationEngine engine) {
 
-        mElapsedTime += frameTime;
+//        mElapsedTime += (frameTime*animationSpeed);
+        mElapsedTime += 1 / 30.0f;
```
is this something that should be hard-coded?

thomasflynn commented on GVRf/Extensions/gvrf-physics/src/main/java/org/gearvrf/physics/GVRPhysicsLoader.java:
```diff
@@ -46,7 +46,8 @@
      * @param scene The scene containing the objects to attach physics components.
      */
-    public static void loadPhysicsFile(GVRContext gvrContext, String fileName, GVRScene scene) {
+    public static void loadPhysicsFile(GVRContext gvrContext, String fileName, GVRScene scene) throws IOException
```
gvr-physicsload no longer compiles because this method now "throws IOException" where it didn't before.  Either this methods needs to not throw IOException, or the demo needs to be updated to catch it.

thomasflynn replied:
it looks like the "throws IOException" is added back in.  does gvr-physicsload still compile?

spodila1 replied:
Yes it is working and the fix was made in the demo not in the framework.

spodila1 commented:
@thomasflynn : I have updated to GVRScene.java and  tested with the demo 
                          gvr-cubemap working

liaxim commented:
If we are to remove the libgdx extension, we should also remove the gvr-modelviewer2 demo app. Ok?

thomasflynn commented:
yes, i think we'll have to do that until it can be updated.

liaxim commented:
gvr-morph is again much slower to load the model. I fixed that master but it doesn't look like a bad merge problem. It is something new.

liaxim commented:
While gvr-widgetviewer shows the integration of libgdx, it also loads and renders gltf models. So the fact that it shows all black models means something is broken there. I guess we can merge as-is if that is what you want.

thomasflynn commented:
we should probably take a look before merging.  which models are coming up black?

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRAssetLoader.java:
```diff
@@ -431,6 +431,10 @@ private void generateLoadEvent()
             }
+            synchronized (mNumTextures)
+            {
+                mNumTextures = -1;
```
Why -1 and not 0. What is the difference?

liaxim replied:
Still curious to know - what is the meaning of -1 here?

NolaDonato replied:
This prevents it from trying to load textures after the asset has been loaded. Theoretically it should never get to generateLoadEvent() unless mTextures is 0.  But it will come here again if it tries to load another texture for that asset. It is basically a guard against having the asset generate a load event twice.

thomasflynn replied:
@spodila1 please include Nola's note above as comments for this line in your next patch.  thanks.

spodila1 replied:
I have included.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMeshMorph.java:
```diff
@@ -320,6 +317,7 @@ public boolean update()
             blendshapeTex.setImage(blendshapeImage);
             mtl.setTexture("blendshapeTexture", blendshapeTex);
         }
+        Log.v("MORPH", "Blend Shape Diffs\n%s", dumpDiffs(20));
```
This log must be removed, right?

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRScene.java:
```diff
@@ -156,17 +156,27 @@ public int removeSceneObjectsByName(final String name) {
         }
 
-        if (null != head) {
+        mSceneRoot = new GVRSceneObject(getGVRContext());
```
What is this change for? I think it might reverting a fix in master.

thomasflynn replied:
@NolaDonato i believe 'git blame' indicated that you changed this code.  do you have insight on the change?

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimation.java:
```diff
@@ -315,11 +350,32 @@ public GVRAnimation setOnFinish(GVROnFinish callback) {
+
+    protected void onFinish()
+    {
+        if (sDebug)
```
Make sDebug final so the compiler can optimize away the branch.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimator.java:
```diff
@@ -202,6 +218,27 @@ public void setRepeatMode(int repeatMode)
+            anim.setOffset(startOffset);
+        }
+    }
+    public void setSpeed(float speed)
```
Javadocs?

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRAnimationEngine.java:
```diff
@@ -114,6 +114,7 @@ public GVRAnimation start(GVRAnimation animation) {
             animation.reset();
             mAnimations.add(animation);
         }
+        animation.onStart();
```
What is onStart for? Doesn't seem to be used for anything.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/GVRSkeleton.java:
```diff
@@ -90,12 +113,14 @@
     final private Quaternionf mTempQuatB = new Quaternionf();
     final private Matrix4f mTempMtx = new Matrix4f();
+    private static int[] sTempBoneParents;
+    private static GVRSceneObject sTempRoot;
```
Static references to GVRf objects are not a good idea in general. In particular the scene object referenced will never be garbage collected. So it is a memory leak.

liaxim replied:
sTempRoot still around. 

Can be changed to
```
        super(root.getGVRContext(), 0);
        setNative(NativeSkeleton.ctor(makeParentBoneIds(root, boneNames)));
```

Will allow makeParentBoneIds to be instance method and sTempRoot to be removed.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/animation/keyframe/BVHImporter.java:
```diff
@@ -250,70 +393,130 @@ public GVRSkeletonAnimation readMotion(BufferedReader buffreader, GVRSkeleton sk
                 bindpose.getLocalRotation(boneIndex, b);
                 q.mul(b);
-                int f = (boneIndex * 5) * frameIndex;
+                f = 5 * frameIndex;
```
A lot of hard-coded numbers in this method. Hard to read and maintain.

liaxim commented:
There are still a ton of whitespace-only/formatting changes to jassimp.cpp which should be reverted.

I still don't get while asset loading APIs are added to the GVRAvatar. They might be avatar-specific but why do we have a GVRAssetLoader in first place if not to collect all apis in one place.

liaxim commented:
gvr-widgetviewer: the initial model is ok; the rest are black.

thomasflynn commented:
i can confirm that the widgetviewer models are black.  and changing the light to a different color only seems to help on the monkey.  so, something is messed up there.
i haven't seen a speed difference in gvr-morph though.  seems to come up in ~2.x seconds on both master and this branch.

liaxim commented:
gvr-morph startup is fixed in the last couple of commits. The logging was slowing it down. If I have to guess, it was ``Log.v("MORPH", "Blend Shape Diffs\n%s", dumpDiffs(20));``.

thomasflynn commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRMeshMorph.java:
```diff
@@ -253,6 +250,7 @@ public void setBlendShape(int index, GVRSceneObject obj)
     public void setBlendShape(int index, GVRVertexBuffer vbuf)
     {
         int shapeDescriptorFlags = 0;
+        float[] vec3dataBase;
```
perhaps i'm going blind, where is this vec3dataBase being used in this method?

NolaDonato replied:
I took this out

liaxim replied:
vec3dataBase seems to be still here.

NolaDonato commented:
I squashed this PR and removed the change made to normalmap.vsh. This fixed the monkey and made the PR more readable.

thomasflynn commented:
ok, the models in the widgetviewer now display correctly.

thomasflynn commented:
there's still a lot of unnecessary white-space changes, but it is apparently difficult to revert them.  i've already given a lecture about it.  is there anything else that needs work in this pull-request?

liaxim commented:
There is still the change to GVRScene.removeAllSceneObjects that reverts a fix from master.

liaxim commented:
To revert a file to a previous version: ``git checkout c744982d38f0acfee5b6b37f9fafb40cae7572ee  ./framework/src/main/java/org/gearvrf/GVRScene.java``. This will make GVRScene go back to the version prior to the changes in this pull request.

liaxim commented:
Same revert operation can be done for jassimp.cpp. The only change there is commented out code removal. "diff -w" ignores whitespace changes.

``git diff -w HEAD upstream/master ./framework/src/main/jni/contrib/jassimp/jassimp.cpp``

```
--- a/GVRf/Framework/framework/src/main/jni/contrib/jassimp/jassimp.cpp
+++ b/GVRf/Framework/framework/src/main/jni/contrib/jassimp/jassimp.cpp
@@ -1454,9 +1454,14 @@ static bool loadMaterials(JNIEnv *env, const aiScene* cScene, jobject& jScene)
 
                for (unsigned int p = 0; p < cMaterial->mNumProperties; p++)
                {
+                       //printf("%s - %u - %u\n", cScene->mMaterials[m]->mProperties[p]->mKey.C_Str(), 
+                       //      cScene->mMaterials[m]->mProperties[p]->mSemantic,
+                       //      cScene->mMaterials[m]->mProperties[p]->mDataLength);
+
                        const aiMaterialProperty* cProperty = cMaterial->mProperties[p];
 
                        lprintf("   converting property %s ...\n", cProperty->mKey.C_Str());
+
                        jobject jProperty = NULL;
                        SmartLocalRef refProperty(env, jProperty);
```

spodila1 commented:
1. I reverted GVRScene.java to the one in master. I am not sure whether to add the following piece of code back. I think it should be added.

`mSceneRoot = new GVRSceneObject(getGVRContext());
        NativeScene.setSceneRoot(getNative(), mSceneRoot.getNative());
        GVRSceneObject parent = head.getParent();
        if (parent != null)
        {
            parent.removeChildObject(head);
        }`
2. Revert operation is done for jassimp.cpp.
3. Removed float[] vec3dataBase;

liaxim commented:
GVRScene.removeAllSceneObjects should be the same as in master unless you intentionally want to change it. It took a while to get it right (see issue https://github.com/Samsung/GearVRf/issues/1962) and if you want to change it somehow - let me know how you verify.

liaxim commented:
Demo apps check out. Last thing is to restore GVRScene.removeAllSceneObjects.

spodila1 commented:
@liaxim restored
